### PR TITLE
Implement automatic cloud proxying

### PR DESCRIPTION
**Pull request by nhooyr**
cc @sreya92
cc @code-asher

Please review carefully. `waitForPath` in particular is very subtle!

### CODE REVIEW

**Comment by nhooyr** [comment]
Only the first commit is relevant, rest are minor CI fixes.

File: src/node/coder-cloud.ts [hunk 1]
```diff
@@ -28,3 +33,125 @@ export async function coderCloudExpose(serverName: string): Promise<void> {
...
+    watchPromise.catch(() => {})
+
+    if (await promisify(fs.exists)(p)) {
+      // The path exists!
+      w.close()
+      return
+    }
+
+    // Now we wait for either the watch promise to resolve/reject or 2000ms.
+    const s = await Promise.race([watchPromise?.then(() => "exists"), delay(2000)])
```
Comment (code-asher) [inline src/node/coder-cloud.ts:151]: Could remove the `?`.

**Review by code-asher — COMMENTED** [review]
`waitForPath` looks solid to me. Do we want to consider Windows support for releasing code-server on Windows or should this feature be OS-gated? I don't know about the agent, but on our end we'd need to modify some of the path handling I think.

File: src/node/coder-cloud.ts [hunk 2]
```diff
@@ -28,3 +33,125 @@ export async function coderCloudExpose(serverName: string): Promise<void> {
     })
   })
 }
+
+export function coderCloudProxy(addr: string) {
+  // addr needs to be in host:port format.
```
[the remark this answers is not in the document]
Comment (nhooyr) [reply src/node/coder-cloud.ts:38]: Nope, the addr passed here is returned from node's http server stuff.

File: src/node/coder-cloud.ts [hunk 3]
```diff
@@ -28,3 +33,125 @@ export async function coderCloudExpose(serverName: string): Promise<void> {
...
+  if (!xdgBasedir.config) {
+    return
+  }
+
+  const sessionTokenPath = path.join(xdgBasedir.config, "coder-cloud", "session")
+
+  const _proxy = async () => {
+    await waitForPath(sessionTokenPath)
+
+    logger.info("exposing coder-server with coder-cloud")
```
[the remark this answers is not in the document]
Comment (nhooyr) [reply src/node/coder-cloud.ts:51]: haha true.

File: src/node/coder-cloud.ts (hunk 1 above)
Comment (nhooyr) [reply src/node/coder-cloud.ts:151]: Yup! nice catch.

File: src/node/coder-cloud.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (code-asher) [reply src/node/coder-cloud.ts:38]: Yup but if you run `code-server --socket /tmp/socket.sock` for example then `addr` here will end up being `/tmp/socket.sock`.

[the remark this answers is not in the document]
Comment (nhooyr) [reply src/node/coder-cloud.ts:38]: Ohhh, sry misread. So I fixed this in the latest code by always listening on `localhost:0` when `--coder-bind` is passed.

**Pull request merged**